File: 14.longest-common-prefix.c
/*
 * @lc app=leetcode id=14 lang=c
 *
 * [14] Longest Common Prefix
 */

char* longestCommonPrefix(char** strs, int strsSize)
{
	if (!strsSize)
	{
		return "";
	}
	if (strsSize == 1)
	{
		return strs[0];
	}
	for (int j = 0;; ++j)
		for (int i = 0; i < strsSize - 1; ++i)
		{
			if (strs[i][j] == '\0')
				return strs[i];
			if (strs[i][j] != strs[i + 1][j])
			{
				strs[0][j] = '\0';
				return strs[0];
			}
		}
}

File: 14.longest-common-prefix.java
/*
 * @lc app=leetcode id=14 lang=java
 *
 * [14] Longest Common Prefix
 *
 * https://leetcode.com/problems/longest-common-prefix/description/
 *
 * algorithms
 * Easy (34.16%)
 * Likes:    2197
 * Dislikes: 1728
 * Total Accepted:    680.3K
 * Total Submissions: 2M
 * Testcase Example:  '["flower","flow","flight"]'
 *
 * Write a function to find the longest common prefix string amongst an array
 * of strings.
 * 
 * If there is no common prefix, return an empty string "".
 * 
 * Example 1:
 * 
 * 
 * Input: ["flower","flow","flight"]
 * Output: "fl"
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: ["dog","racecar","car"]
 * Output: ""
 * Explanation: There is no common prefix among the input strings.
 * 
 * 
 * Note:
 * 
 * All given inputs are in lowercase letters a-z.
 * 
 */

// @lc code=start
class Solution {
    public String longestCommonPrefix(String[] strs) {
        if (strs.length == 0)
            return "";
        else if (strs.length == 1)
            return strs[0];
        for (int i = 0; true; i++) {
            for (int j = 0; j < strs.length; j++) {
                if (i >= strs[j].length())
                    return strs[j];
                if (strs[j].charAt(i) != strs[0].charAt(i)) {
                    return strs[0].substring(0, i);
                }
            }
        }
    }
}
// @lc code=end
